mewarpx/mewarpx/emission.py

```diff
@@ -0,0 +1,867 @@
+        self.weight = weight
+        self.rseed = rseed
+        self.name = name
+        if self.name is None:
+            self.name = "fixed_injector_" + self.species.name
+        self.unique_particles = unique_particles
+
+        print(("Fixed injection of {} particles, weight {}, every "
+            "{} timesteps.").format(
+                self.npart_total, self.weight, self.injectfreq))
+        callbacks.installparticleinjection(self.inject_particles)
+
+    def inject_particles(self):
+        """Perform the actual injection!"""
+        effective_it = mwxrun.get_it() - self.injectoffset
+        if effective_it >= 0 and effective_it % self.injectfreq == 0:
+
+            # Adjust npart for processor number if needed
+            npart = self.compute_npart(
+                npart_total=self.npart_total,
+                unique_particles=self.unique_particles
+            )
+
+            # TODO randomdt and velhalfstep are False simply because they're
+            # not supported at present
+            particles_dict = self.emitter.get_newparticles(
+                npart=npart, w=self.weight,
+                q=self.species.sq, m=self.species.sm,
+                rseed=self.rseed,
+                randomdt=False, velhalfstep=False
+            )
+
+            print("Inject {} particles".format(len(particles_dict['x'])))
+            # We can basically call _libwarpx.libwarpx.warpx_addNParticles()
+            # directly as we shouldn't need the wrapper, unless number to
+            # inject is 0, but for now we'll be conservative and use the
+            # wrapper.
+
+            # Note some parts of WarpX call the variables ux and some parts vx,
+            # and they're referred to as momenta. But I don't see anywhere
+            # they're actually used as momenta including the particle mass -
+            # the actual update is in Source/Particles/Pusher/UpdatePosition.H
+            _libwarpx.add_particles(
+                species_number=self.species.species_number,
+                x=particles_dict['x'],
+                y=particles_dict['y'],
+                z=particles_dict['z'],
+                ux=particles_dict['vx'],
+                uy=particles_dict['vy'],
+                uz=particles_dict['vz'],
+                # E_total=particles_dict['E_total'],
+                attr=particles_dict['w'],
+                unique_particles=self.unique_particles
+            )
+
+            if self.injector_diag is not None:
+                self.record_injectedparticles(
+                    self.species,
+                    particles_dict['E_total'],
+                    particles_dict.get('w', None)
+                )
+
+    def cleanup(self):
+        """Uninstall particle injection function
+        Previously used for testing to put Warp into a good state; could also
+        used if simulation should change partway through.
+        """
+        callbacks.uninstallparticleinjection(self.inject_particles)
```

Comment:
Let's delete this whole function. We've never used it which means it's not going to get CI-tested etc. Easy to write if we ever do need it.